broker/value_added/classify.py

```diff
@@ -0,0 +1,91 @@
+                                        (mjd, flux, fluxerr, passband,
+                                        photflag, ra, dec, objid, redshift, mwebv)
+
+        cx_dicts        (list of dicts): candidate and xmatch info as needed for
+                                        BigQuery classification table.
+        Note:
+            A given index must identify the same unique object + host galaxy combination
+            in both the light_curves and cx_dicts lists.
+            Use the function value_added.format_for_rapid() to get both lists.
+
+    Returns:
+        List of dictionaries formatted for upload to BigQuery classification table.
+    """
+
+    classification = Classify(known_redshift=use_redshift)
+    predictions = classification.get_predictions(light_curves, \
+                    return_predictions_at_obstime=True, return_objids=True)
+
+    if plot:
+        # Plot classifications vs time
+        classification.plot_light_curves_and_classifications(figdir=plotdir)
+        # classification.plot_classification_animation(figdir=plotdir)
+        print('\nRAPID classification plot(s) saved to {}'.format(plotdir))
+        print("Pass arg 'plotdir' to change this location.")
+
+    class_dicts_list = format_rapid_for_BQ(cx_dicts, predictions)
+
+    return class_dicts_list
+
+
+def format_rapid_for_BQ(cx_dicts, predictions):
+    """ Combines RAPID output with candidate and xmatch data and formats for upload to BigQuery.
+
+    Args:
+        cx_dicts       (dict of dicts): candidate and xmatch info as needed for
+                                        BigQuery classification table.
+                                        Keys are lightcurve IDs (cxid, as given by
+                                        value_added.alert_xobj_id()).
+
+        predictions            (tuple): as returned by rapid function get_predictions()
+
+    Returns:
+        cx_dicts with candidate epoch classification info added to each dictionary.
+    """
+
+    # classes_list, times_list, objids_list = predictions # lists of arrays
+
+    # Check that all light curves were classified --- no longer needed, have cxid
+    # err_msg = "Number of predictions from RAPID != Number of input light curves. \
+    #             \nCannot match predictions to candid's."
+    # assert len(classes_list) == len(cx_dicts), err_msg
+
+    # Merge predicitions with alert and hostgal info
+    class_dicts_list = []
+    for lc_preds, lc_times, cxid in zip(*predictions):
+        # Get the predictions for the current alert candidate only
+        cand_preds = lc_preds[-1] # check this.
+        # I tried to match the candidate mjd (see below), but the output times are
+        # not exactly the same as the input times.
+        class_keys = ['prob_class{}'.format(i) for i in range(len(cand_preds))] # fix this.
+        preds_dict = dict(zip(class_keys, cand_preds.tolist()))
+        # # find position of cand_mjd in lc_times array
```

Comment:
I'm not quite following the comments here. It seems like you were struggling to figure this part out. If it helps, we can chat for a bit in person and build this out.

Reply:
I was, the input times are interpolated for the classification and so the outputs are not exactly at the input times. To get this running, it didn't really matter which time I chose to plot the classification, so I finally just picked what I'm pretty sure is the right answer anyway (the last one). Let's chat.